### DISCUSSION TITLE
Not able to get tags value in readings with EdgeX rules engine

### DISCUSSION THREAD

**Discussion by ymcapig** [lf-edge/ekuiper#3139] [Q&A]
Hello folks,
I have run the eKuiper rule engine with EdgeX. I followed 3.1Napa documentation to create an EdgeX stream and a rule as follows.
#rules
  SELECT meta(*),* FROM Demo

#device profile 
...
deviceResources:
       -description: Test
         name: Test
         ...
         tags:
            tag1:value
            tag2:value
            tag3:value
         ...
[Question]
I set the tags value in device resource as above, and could saw the tag1/tag2/tag3 values in readings with EdgeX metadata.
But I can't get these tags value in Redis where EdgeX rules engine sunk to. 
Could any way to get tags values in readings by rules engine?
It would be appreciated if anyone could share best practices. 
Thanks

**Comment by ngjaying** [comment]
What is your full rule? It should contain definition of "actions"

**Reply by ymcapig** [reply]
My full rule:
  "sql": "SELECT meta(*),* FROM demo",
  "actions": [
    {
      "edgex": {
        "type": "redis",
        "protocol": "redis",
        "host": "edgex-redis",
        "port": 6379,
        "contentType": "application/json",
        "messageType": "event",
        "sendSingle": false,
        "dataTemplate": "",
        "deviceName": "",
        "metadata": "",
        "optional": {
          "Qos": "0",
          "Retained": "false",
          "SkipCertVerify": "true",
          "AutoReconnect": "true",
          "CleanSession": "true"
        },
        "profileName": "",
        "topic": "demo/edgex",
        "topicPrefix": ""
      }
    }
  ],
  "options": {
    "isEventTime": false,
    "sendMetaToSink": false,
    "sendError": true,
    "qos": 0,
    "debug": false,
    "logFilename": "",
    "lateTolerance": 1000,
    "concurrency": 1,
    "bufferLength": 1024,
    "checkpointInterval": 300000,
    "restartStrategy": {
      "attempts": 0,
      "delay": 1000,
      "multiplier": 2,
      "maxDelay": 30000,
      "jitter": 0.1
    },
    "cron": "",
    "duration": "",
    "cronDatetimeRange": null
  }

**Comment by ngjaying** [comment]
Check this doc https://ekuiper.org/docs/en/latest/edgex/edgex_rule_engine_command.html and other edgex related doc.

**Reply by ymcapig** [reply]
Thanks. 
And I checked the code of ekuiper source.go [[source](https://github.com/lf-edge/ekuiper/blob/6743f39170be3e48bbdb512537b22ae2fae356b3/internal/io/edgex/source.go#L190)]
<img width="926" alt="image" src="https://github.com/user-attachments/assets/3019fe85-a4af-4096-81f0-29901c79b4c9">
Is that means rules engine doesn't get tags value in readings?
The reading:
{"id":"49cd4c3b-ac5b-4616-b1d6-ce82ba19fdd6","origin":1724403812816031200,"deviceName":"69176510-db76-4470-8141-13c09f0a50f1","resourceName":"abc","profileName":"824cbfd2-4998-4a1d-8bec-488ff82800cf","valueType":"Uint16","tags":{"TAG"}
